File: main/IDE/convert.js
function isVariable(el) {
    return el.trim().match(/[_a-z][_a-z0-9]*/i)
}
function getParams(el) {
    let eltab = el.split(',').filter((str) => str != '')
    let alls = [], unchanges = [], changes = []
    for (v in eltab) {
        i = eltab[v]
        i = i.replace(/ +/i, ' ')
        if (i == ' ' || i == '') { continue }
        test = i.match(/(?<name>^@?[a-z_]+([a-z0-9_])*)(:[a-z_]+([a-z0-9_])*)?$/i)

        if (test && test[1]) {
            if (test[1][0] != '@') {
                unchanges.push(["__OLD" + test[1], test[1]])
                alls.push(test[1])
            } else {
                changes.push(test[1].substring(1))
                alls.push(test[1].substring(1))
            }
        } else {
            return [false, false, false]
        }
    }
    return [alls, unchanges, changes]
}
function isBoucleFor(el) {
    el = el.replace(/[ ]+/, ' ')
    el = el.split(" ").slice(1, -1)
    var index, start, end = 0
    for (let i = 0; i < el.length; i++) {
        if (el[i].trim() == "de") { index = i }
        if (el[i].trim() == "a") { start = i }
    }

    return `for (let ${el.slice(0, index)}=${el.slice(index + 1, start).join(" ")};i<${el.slice(start + 1, el.length).join(" ")}),i++){`
}
function isSi(el) {
    let newel = el.match(/si[ ]+(?<arguments>.+)?[ ]+alors:?/i)
    if (newel) {
        return `if (${newel[1]}){`
    } else {
        return false
    }
}
function isSinonsi(el) {
    let newel = el.match(/sinonsi[ ]+(?<arguments>.+)?[ ]+alors:?/i)
    if (newel) {
        return `}else if (${newel[1]}){ `
    } else {
        return false
    }
}
function isSinon(el) {
    var newel = el.match(/sinon[ ]*:?/i)
    if (newel) {
        return `}else{ `
    } else {
        return false
    }
}
function isFonction(el) {
    let newel = el.match(/fonction[ ]+(?<name>[a-z]([a-z0-9_])*)[ ]*\((?<arguments>.*)\)[ ]*:[ ]*(?<returntype>[a-z]+[0-9]*)/i)
    if (newel) {
        let [alls, unchanges, changes] = getParams(newel[3])
        if (alls == []) {
            return [`function ${newel[1]} ()/*${newel[4]}*/{`, [], []]
        }
        if (unchanges || changes) {
            return [`function ${newel[1]} (${alls.join(',')}){/*${newel[4]}*/`, unchanges, changes]
        }
    } else {
        return [false, false, false]
    }
    return [false, false, false]
}
function isProcedure(el) {
    newel = el.match(/procedure[ ]+(?<name>[a-z]([a-z0-9_])*)[ ]*\((?<arguments>.*)\)[ ]*:?/i)
    if (newel) {
        let [alls, unchanges, changes] = getParams(newel[3])
        console.log(alls, unchanges, changes)
        if (alls == []) {
            return [`procedure ${newel[1]} (){`, [], []]
        }
        if (unchanges || changes) {
            return [`procedure ${newel[1]} (${alls.join(',')}){`, unchanges, changes]
        }
    } else {
        return [false, false, false]
    }
    return [false, false, false]
}
function isTantque(el) {
    newel = el.match(/tantque[ ]+(?<arguments>.*)/i)
    if (newel) {
        return `while (${newel[1]}){ `
    } else {
        return False
    }
}
function isJusqua(el) {
    newel = el.match(/Jusqu'?a[ ]+(?<arguments>.+)/i)
    if (newel) {
        return `if (${newel[1]}){ `
    } else { return False }
}
function getSpanTable(start, enD) {
    let begin = Number(start.charCodeAt(0))
    let ender = Number(enD.charCodeAt(0))
    let tab = []
    for (let i = begin; i < ender + 1; i++) {
        tab.push(String.fromCharCode(i))
    }
    return tab
}
function replacement(el) {
    return L = el.replace(/\[A\.\.Z\]/, `${getSpanTable('A', 'Z')}`)
        .replace(/\[a\.\.z\]/, `${getSpanTable('a', 'z')}`)
        .replace(/\[0\.\.9\]/, `${getSpanTable('0', '9')}`)
        .replace(/[^a-z0-9_]non[^a-z0-9_]/i, " ! ")
        .replace(/[^a-z0-9_]et[^a-z0-9_] /i, " && ")
        .replace(/[^a-z0-9_]ou[^a-z0-9_] /i, " || ")
        .replace(/[^a-z0-9_]dans[^a-z0-9_]/i, " in ")
        .replace(/ +/, " ")
        .replace(/=/, "==")
        .replace(/====/, "==")
        .replace(/ div /, " // ")
        .replace(/ mod /, " % ")
        .replace(/\^/, " ** ")
        .replace(/<--/, '=')

}

function replaceInString(el) {
    let opened = null
    let start = 0
    let end = el.length
    let quoted = []
    let legit = []
    let full = ''
    for (let i = 0; i < el.length; i++) {
        const item = el[i].trim();
        let before; if (i > 0) { before = i - 1 }
        if ((item == '"' || item == "'") && el[before] != '\\') {
            if (opened == null) {
                opened = i, end = i - 1
                full += replacement(el.substring(start, end + 1))
            } else if (el[opened] == item) {
                quoted.push([opened, i])
                legit.push([start, end])
                full += el.substring(opened, i + 1)
                start = i + 1, end = el.length - 1, opened = null
            }
        }
    }
    legit.push([start, end])
    full += el.substring(start, end + 1)
    if (opened != null) {
        console.log('unformatted string')
    }
    console.log(full)
    return full
}
function translateLines(res) {
    newres = []
    for (const [i, v] of Object.entries(res)) {
        key = Object.keys(res[i])
        console.log(key, res[i][key])
        res[i][key] = replaceInString(res[i][key])
        console.log(key, res[i][key])
        // res[i][key] = res[i][key].replace(/([^:]*)(:+)$/, /$1/)
        res[i][key] = res[i][key].replace(/^(selon) (.+)/, "match $2:")
        let starter = res[i][key].split(' ')[0].toLowerCase().trim()
        if (re.match("^pour[ ]+", starter, re.IGNORECASE)) {
            test = isBoucleFor(res[i][key])
            if (test) {
                newres.append(test + '{/*' + str(key) + '*/') 
            } else {
                console.error("pour invalide")
            }
        }
    }
    console.log(newres)
    return newres
}

File: src/convert.py
import re 
import syntax as stx

def getSpanTable(start, ende):
    begin = int(ord(start))
    ender = int(ord(ende))
    tab = []
    for i in range(begin, ender + 1): tab.append(chr(i))
    return tab

def replacement(el):
    el1 = el
    el1 = re.sub("\[A\.\.Z\]", f" {getSpanTable('A','Z')} ", el1)
    el1 = re.sub("\[a\.\.z\]", f" {getSpanTable('a','z')} ", el1)
    el1 = re.sub("\[0\.\.9\]", f" {getSpanTable('0','9')} ", el1)
    el1 = re.sub("[^a-z0-9_]non[^a-z0-9_]", " not ", el1, re.IGNORECASE)
    el1 = re.sub("[^a-z0-9_]et[^a-z0-9_]", " and ", el1, re.IGNORECASE)
    el1 = re.sub("[^a-z0-9_]ou[^a-z0-9_]", " or ", el1, re.IGNORECASE)
    el1 = re.sub("[^a-z0-9_]dans[^a-z0-9_]", " in ", el1, re.IGNORECASE)
    el1 = re.sub(" +", " ", el1)
    el1 = re.sub("=", "==", el1, re.IGNORECASE) 
    el1 = re.sub(" div ", " // ", el1, re.IGNORECASE)
    el1 = re.sub(" mod ", " % ", el1, re.IGNORECASE)
    el1 = re.sub("\^", "**", el1, re.IGNORECASE) 

    return el1
def replaceInString(el):
    opened = None
    start = 0
    end = len(el)
    quoted = []
    legit = []
    full = ""
    for i in range(0, len(el)):
        item = el[i].strip()
        before = 0
        if i > 0: before = i - 1 

        if (item == '"' or item == "'") and el[before] != "\\":
            if opened == None:
                opened = i
                end = i - 1
                full += replacement(el[start : end + 1])
            elif el[opened] == item:
                quoted.append([opened, i])
                legit.append([start, end])
                full += el[opened : i + 1]
                start = i + 1
                end = len(el) - 1
                opened = None
    legit.append([start, end])
    full += replacement(el[start : end + 1])
    if opened != None: exit("unformatted String ")

    return full

def read(lines,start,end):
    jdid = []
    for i in range(start+1,end-1): 
        line = lines[i].strip()
        line = replaceInString(line)
        line = re.sub("([^:]*)(:+)$", r"\1", line)
        starter =  line.split(" ")[0].lower().strip() + " "

        if re.match("(.*)[<\--](.*)", line, re.IGNORECASE):
            print('ggggg')
            test = stx.assignment(line)
            if test : jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else    : exit(f"assignment invalide line {i}")

        elif re.match("^pour[ ]+", starter, re.IGNORECASE):
            test = stx.pour(line)
            if test : jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else    : exit(f"pour invalide line {i} ")

        elif re.match("^si[ ]+", starter, re.IGNORECASE):
            test = stx.si(line)
            if test: jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else: exit(f"si invalide line {i} ")

        elif re.match("^sinonsi[ ]+", starter, re.IGNORECASE):
            test = stx.sinonsi(line)
            if test:jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else: exit(f"sinonsi invalide line {i} ")

        elif re.match("^sinon[ ]*:?", starter, re.IGNORECASE):
            test = stx.sinon(line)
            if test:jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else: exit(f"sinon invalide line {i} ")

        elif re.match("^tant[ ]?que", starter, re.IGNORECASE):
            test = stx.tantque(line)
            if test:jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else: exit(f"tantque invalide line {i} ")

        elif re.match("^jusqu'?(a|à)", starter, re.IGNORECASE):
            test = stx.jusqua(line)
            if test:
                jdid.append(f'{test}#{str(len(jdid))}\n\tbreak#{str(len(jdid)+1)}')
                jdid.append("finsi" + "#" + str(len(jdid)))
            else: exit(f"Jusqu'a invalide line {i} ")

        elif re.match("^R(e|é)p(e|é)ter[ ]*", starter, re.IGNORECASE):
            test = stx.repeter(line)
            if test: jdid.append(f'{test}#{str(len(jdid))} \n\tpass#{str(len(jdid)+1)}')
            else: exit(f"repeter invalide line {i} ")
        else : jdid.append(line)
    print(jdid)
    return jdid